### DISCUSSION TITLE
How does Nushell embeds CoreUtils into its binary?

### DISCUSSION THREAD

**Discussion by ClementNerma** [nushell/nushell#13100] [Q&A] [answered]
When using some functions like `cp` for instance, Nushell uses `uutils/coreutils` under the hood to get GNU-like behaviour on Windows.

But I'm wondering how it's embedded? There is no documentation on coreutils on how to embed these into the binary. I suppose there Nushell has a way to call the embedded coreutils program when it's called a certain way?

I'm currently building a shell so I'm genuinely curious about this question :)

**Comment by fdncred** [comment, marked as answer]
uutils/coreutils publish all their commands as crates. so we embed uu_cp, uu_mv, and others. however, there was a lot of work that went into those so that we could use them. settings had to be exposed from the uutils side so we can keep producing nice errors when a uu command fails. you can search our PRs by the coreutils label and find how they were impelmented.

**Reply by ClementNerma** [reply]
Ok so it's not trivial at all to integrate coreutils into another library currently.
Thanks for your answer :)

**Reply by youk** [reply]
If built-in `cp` doesn't support a particular flag implemented in uutils/coreutils (`--attributes-only`), does it make sense to request it? Or unlikely to happen upon request?

**Reply by fdncred** [reply]
We choose the parameters we think are most useful. That doesn't mean we always make the right choices though. Feel free to make your case in an issue or our discord #implementation-chat channel. We definitely have the ability to change them if we agree that it's a good move for the community.

**Reply by youk** [reply]
Got it, thanks!
